hive/src/main/java/org/apache/iceberg/hive/legacy/HiveExpressions.java

```diff
@@ -0,0 +1,271 @@
+import org.apache.iceberg.expressions.BoundPredicate;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.ExpressionVisitors;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.expressions.Literal;
+import org.apache.iceberg.expressions.UnboundPredicate;
+import org.apache.iceberg.expressions.UnboundTerm;
+
+
+class HiveExpressions {
+
+  private HiveExpressions() {}
+
+  /**
+   * Simplifies the {@link Expression} so that it fits the restrictions of the expression that can be passed
+   * to the Hive metastore. For details about the simplification, please see {@link Simplify}
+   * @param expr The {@link Expression} to be simplified
+   * @param partitionColumnNames The set of partition column names
+   * @return TRUE if the simplified expression results in an always true expression or if there are no predicates on
+   *            partition columns in the simplified expression,
+   *         FALSE if the simplified expression results in an always false expression,
+   *         otherwise returns the simplified expression
+   */
+  static Expression simplifyPartitionFilter(Expression expr, Set<String> partitionColumnNames) {
+    Expression simplified = ExpressionVisitors.visit(expr, new Simplify(partitionColumnNames));
+    if (simplified != null) {
+      // During simplification of IN, NOTIN, NULL, and NOT NULL expressions we introduce additional NOT, TRUE, and FALSE
+      // expressions; so we call Simplify again to remove them
+      simplified = ExpressionVisitors.visit(simplified, new Simplify(partitionColumnNames));
+    }
+    return (simplified == null) ? Expressions.alwaysTrue() : simplified;
+  }
+
+  /**
+   * Converts an {@link Expression} into a filter string which can be passed to the Hive metastore
+   * @param expr The {@link Expression} to be converted into a filter string. This expression must fit the restrictions
+   *             on Hive metastore partition filters. For more details, see {@link Simplify}
+   * @return a filter string equivalent to the given {@link Expression} which can be passed to the Hive metastore
+   */
+  static String toPartitionFilterString(Expression expr) {
+    return ExpressionVisitors.visit(expr, ExpressionToPartitionFilterString.get());
+  }
+
+  /**
+   * Simplifies the {@link Expression} so that it fits the restrictions of the expression that can be passed
+   * to the Hive metastore. It performs the following changes:
+   * 1. Removes any predicates and non-partition columns
+   * 2. Rewrites NOT operators
+   * 3. Removes IS NULL and IS NOT NULL predicates (Replaced with FALSE and TRUE respectively as partition column values
+   *    are always non null for Hive)
+   * 4. Expands IN and NOT IN operators into ORs of EQUAL operations and ANDs of NOT EQUAL operations respectively
+   * 5. Removes any children TRUE and FALSE expressions (Note that the simplified expression still can be TRUE and FALSE
+   *    at the root)
```

Comment:
Done.

Starts_with is the only predicate that we do not handle at this moment. I don't see any usage of it inside LI. I think if we see people using it we can rewrite the filter as alwaysTrue and not fail, but I would rather fail and let users change it to a combination of `>` and `<` so that they actually do partition pruning.